### PR TITLE
issue #675 enable tenant/datastore support for searchalltest

### PR DESCRIPTION
**Pull request by albertwang-ibm**
Signed-off-by: Albert Wang <xuwang@us.ibm.com>

### CODE REVIEW

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
+/*
+ * (C) Copyright IBM Corp. 2019
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: date
```suggestion
 * (C) Copyright IBM Corp. 2019, 2020
```

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java [hunk 2]
```diff
@@ -767,15 +768,16 @@ protected synchronized Client getClient() throws Exception {
             // Add a hostname verifier if we're using an ssl transport.
             if (usingSSLTransport() && !isHostnameVerificationEnabled()) {
                 cb = cb.hostnameVerifier(new HostnameVerifier() {
+                    @Override
```
Comment (prb112) [inline fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: it's not really an override? it's a specific implementation

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java [hunk 3]
```diff
@@ -452,10 +452,11 @@ public FHIRResponse searchAll(FHIRParameters parameters, boolean isPost, FHIRReq
         } else {
             endpoint = endpoint.path("_search");
             builder = endpoint.request(getDefaultMimeType());
+            builder = addRequestHeaders(builder, headers);
```
Comment (prb112) [inline fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: addRequestHeaders - where does this come from what does it do?

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java (hunk 1 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: ha, good catch. will change

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java (hunk 3 above)
Comment (albertwang-ibm) [reply fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: this is used for adding the tenant and datastore headers.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java [hunk 4]
```diff
@@ -390,7 +407,10 @@ public void testCreatePatientAndObservationWithUniqueTag() throws Exception {
                         .build();
 
         Entity<Patient> entity = Entity.entity(patient, FHIRMediaType.APPLICATION_FHIR_JSON);
-        Response response = target.path("Patient").request().post(entity, Response.class);
+        Response response = target.path("Patient").request()
```
Comment (prb112) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java]: Should the test be renamed to indicate that it is testing a different tenant?

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java (hunk 2 above)
Comment (albertwang-ibm) [reply fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: en... interesting... this was not added by me, let me try to remove it and see

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java (hunk 4 above)
Comment (prb112) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java]: maybe event a comment explaining what is happening

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java (hunk 2 above)
Comment (albertwang-ibm) [reply fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: can not be removed, it's always be added back automatically. and the function is abstract, so seems have to override.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java (hunk 4 above)
Comment (albertwang-ibm) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java]: I wanted to just enable it for tenant specific test, but I don't want to change the default behavior - testing the default ds of the default tenant.    I just want to change the tenant in my local to do some simple performance tests against my test tenants.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+               }
+
+            long end = System.nanoTime();
+            logger.info(String.format("Total Resources: %d, Took: %6.3f seconds", totalNum, (end-start)/NANOS));
+        }
+        catch (Exception x) {
+            logger.log(Level.SEVERE, "Failed to run", x);
+        }
+        finally {
+            System.exit(0);
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: I don't think this is necessary... Let it naturally fall out.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java (hunk 4 above)
Comment (albertwang-ibm) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java]: good point, let me add comments

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+                    if (bufferStream.size() > 0) {
+                        dataPackTags.add(BulkDataUtils.multiPartUpload(cosClient, cosBucketName, segName,uploadId,
+                                new ByteArrayInputStream(bufferStream.toByteArray()), bufferStream.size(), partNum++));
+                        bufferStream.reset();
+                    }
+
+                    if (segNum < numberOfFiles - 1 && lineRed == numOfRes4Seg
+                            || !isMore2Read) {
+                        BulkDataUtils.finishMultiPartUpload(cosClient, cosBucketName, segName, uploadId, dataPackTags);
+                        System.out.println("Finished writting for " + segName);
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: Use a logger

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java (hunk 5 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: I added this on purpose for running inside eclipse, without this, in my tests, eclipse always shows the app is not finished - very annoying.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+                System.exit(1);
+            }
+
+            S3Object item = cosClient.getObject(new GetObjectRequest(cosBucketName, cosFile2Break));
+            try (S3ObjectInputStream s3InStream = item.getObjectContent();
+                    BufferedReader resReader = new BufferedReader(new InputStreamReader(s3InStream))) {
+                   totalNum = m.getFhirResourceNumberFromBufferReader(resReader);
+               } catch (Exception ioe) {
+                   logger.warning("Error proccesing file " + cosFile2Break + " - " + ioe.getMessage());
+                   System.exit(1);
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: There are a lot of System.exits here... .throw an exception, and handle one time.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+import com.ibm.cloud.objectstorage.services.s3.AmazonS3;
+import com.ibm.cloud.objectstorage.services.s3.model.GetObjectRequest;
+import com.ibm.cloud.objectstorage.services.s3.model.PartETag;
+import com.ibm.cloud.objectstorage.services.s3.model.S3Object;
+import com.ibm.cloud.objectstorage.services.s3.model.S3ObjectInputStream;
+import com.ibm.fhir.bulkcommon.BulkDataUtils;
+import com.ibm.fhir.bulkcommon.Constants;
+
+/**
+ * Tool to break large COS file into multiple ones.
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: Elaborate as to why it's being done, and what the case for using it is... recommendation, not a requirement of the review.

Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: let me add comments

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java (hunk 7 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: make sense

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java (hunk 8 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java (hunk 7 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java (hunk 6 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: done

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java (hunk 4 above)
Comment (albertwang-ibm) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java (hunk 1 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java]: done

**Pull request merged**